File: swift/test/Ice/facets/TestI.swift
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

import Ice

class AI: A {
    func callA(current _: Ice.Current) throws -> String {
        return "A"
    }
}

class BI: B {
    func callA(current _: Ice.Current) throws -> String {
        return "A"
    }

    func callB(current _: Ice.Current) throws -> String {
        return "B"
    }
}

class CI: C {
    func callA(current _: Ice.Current) throws -> String {
        return "A"
    }

    func callC(current _: Ice.Current) throws -> String {
        return "C"
    }
}

class DI: D {
    func callA(current _: Ice.Current) throws -> String {
        return "A"
    }

    func callB(current _: Ice.Current) throws -> String {
        return "B"
    }

    func callC(current _: Ice.Current) throws -> String {
        return "C"
    }

    func callD(current _: Ice.Current) throws -> String {
        return "D"
    }
}

class EI: E {
    func callE(current _: Ice.Current) throws -> String {
        return "E"
    }
}

class FI: F {
    func callE(current _: Ice.Current) throws -> String {
        return "E"
    }

    func callF(current _: Ice.Current) throws -> String {
        return "F"
    }
}

class GI: G {
    var _communicator: Ice.Communicator

    public init(communicator: Ice.Communicator) {
        _communicator = communicator
    }

    func callG(current _: Ice.Current) throws -> String {
        return "G"
    }

    func shutdown(current _: Ice.Current) throws {
        _communicator.shutdown()
    }
}

class HI: H {
    let _communicator: Ice.Communicator

    init(communicator: Ice.Communicator) {
        _communicator = communicator
    }

    func callG(current _: Ice.Current) throws -> String {
        return "G"
    }

    func callH(current _: Ice.Current) throws -> String {
        return "H"
    }

    func shutdown(current _: Ice.Current) throws {
        _communicator.shutdown()
    }
}

File: python/test/Ice/admin/TestI.py
#
# Copyright (c) ZeroC, Inc. All rights reserved.
#

import Ice
import Test
import threading


def test(b):
    if not b:
        raise RuntimeError("test assertion failed")


class TestFacetI(Test.TestFacet):
    def op(self, current):
        return


class RemoteCommunicatorI(Test.RemoteCommunicator, Ice.PropertiesAdminUpdateCallback):
    def __init__(self, communicator):
        self.communicator = communicator
        self.called = False
        self.m = threading.Condition()

    def getAdmin(self, current):
        return self.communicator.getAdmin()

    def getChanges(self, current):
        with self.m:
            #
            # The client calls PropertiesAdmin::setProperties() and then invokes
            # this operation. Since setProperties() is implemented using AMD, the
            # client might receive its reply and then call getChanges() before our
            # updated() method is called. We block here to ensure that updated()
            # gets called before we return the most recent set of changes.
            #
            while not self.called:
                self.m.wait()

            self.called = False

            return self.changes

    def shutdown(self, current):
        self.communicator.shutdown()

    def waitForShutdown(self, current):
        #
        # Note that we are executing in a thread of the *main* communicator,
        # not the one that is being shut down.
        #
        self.communicator.waitForShutdown()

    def destroy(self, current):
        self.communicator.destroy()

    def updated(self, changes):
        with self.m:
            self.changes = changes
            self.called = True
            self.m.notify()


class RemoteCommunicatorFactoryI(Test.RemoteCommunicatorFactory):
    def createCommunicator(self, props, current):
        #
        # Prepare the property set using the given properties.
        #
        init = Ice.InitializationData()
        init.properties = Ice.createProperties()
        for k, v in props.items():
            init.properties.setProperty(k, v)

        #
        # Initialize a new communicator.
        #
        communicator = Ice.initialize(init)

        #
        # Install a custom admin facet.
        #
        communicator.addAdminFacet(TestFacetI(), "TestFacet")

        #
        # The RemoteCommunicator servant also implements PropertiesAdminUpdateCallback.
        # Set the callback on the admin facet.
        #
        servant = RemoteCommunicatorI(communicator)
        admin = communicator.findAdminFacet("Properties")
        if admin is not None:
            admin.addUpdateCallback(servant)

        proxy = current.adapter.addWithUUID(servant)
        return Test.RemoteCommunicatorPrx.uncheckedCast(proxy)

    def shutdown(self, current):
        current.adapter.getCommunicator().shutdown()
